# 🚀 머신러닝 실습 : 고객 구매 데이터로 성별 예측 모델링 (분류 문제)

* 주어진 데이터는 백화점 고객의 1년 간 구매 데이터입니다.
* 고객 3,500명에 대한 학습용 데이터(y.csv, X.csv)를 이용하여 성별예측 모형을 만들어보세요.
* 모델의 성능은 자유롭게 측정해봅니다!

## [실습 프로세스]
1. 데이터 불러오기  
2. 데이터 탐색
3. 데이터 전처리  
4. 학습/테스트 데이터 분리  
5. 모델 선택 및 학습  
6. 예측 및 평가  


<br/>

---

<br/>
<br/>

# 0. 라이브러리 불러오기

* 라이브러리를 가져와서 과정을 준비합니다

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

import warnings
warnings.filterwarnings(action='ignore')

<br/>

---

<br/>
<br/>

# 1. 데이터 불러오기
* 데이터를 가져와서 과정을 준비합시다.
- 인코딩 방식은 'euc-kr' 을 활용하세요.
- 데이터 출처 : 한국데이터산업진흥원 빅데이터분석기사 실기 공개 예시 문항

- 독립 변수 데이터셋 : ./data/X.csv
- 종속 변수 데이터셋 : ./data/y.csv


데이터 파일을 불러옵니다. 보통 CSV 파일을 pandas로 읽어옵니다.

In [2]:
import os
# 노트북 파일이 있는 폴더로 이동 (예시)
os.chdir(r'C:\githome\hipython_rep')

# 변경 후 확인
print("변경 후:", os.getcwd())

변경 후: c:\githome\hipython_rep


In [3]:
X = pd.read_csv('./data/X.csv', encoding='euc-kr')
y = pd.read_csv('./data/y.csv', encoding='euc-kr')

<br/>

---

<br/>
<br/>

# 2. 데이터 탐색하기
* 데이터를 이해할 수 있도록 탐색과정을 수행해봅시다.


데이터의 상위 몇 개 행을 출력하여 전체 구조를 미리 확인합니다.

In [4]:
print(X.head())  
print(y.head())  


   cust_id      총구매액     최대구매액       환불금액   주구매상품 주구매지점  내점일수   내점당구매건수  \
0        0  68282840  11264000  6860000.0      기타   강남점    19  3.894737   
1        1   2136000   2136000   300000.0     스포츠   잠실점     2  1.500000   
2        2   3197000   1639000        NaN  남성 캐주얼   관악점     2  2.000000   
3        3  16077620   4935000        NaN      기타   광주점    18  2.444444   
4        4  29050000  24000000        NaN      보석  본  점     2  1.500000   

     주말방문비율  구매주기  
0  0.527027    17  
1  0.000000     1  
2  0.000000     1  
3  0.318182    16  
4  0.000000    85  
   cust_id  gender
0        0       0
1        1       0
2        2       1
3        3       1
4        4       0



데이터의 요약 정보나 통계 정보를 출력해 변수들의 유형과 분포를 확인합니다.

In [5]:
print(X.info())  
print(y.info())


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3500 entries, 0 to 3499
Data columns (total 10 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   cust_id  3500 non-null   int64  
 1   총구매액     3500 non-null   int64  
 2   최대구매액    3500 non-null   int64  
 3   환불금액     1205 non-null   float64
 4   주구매상품    3500 non-null   object 
 5   주구매지점    3500 non-null   object 
 6   내점일수     3500 non-null   int64  
 7   내점당구매건수  3500 non-null   float64
 8   주말방문비율   3500 non-null   float64
 9   구매주기     3500 non-null   int64  
dtypes: float64(3), int64(5), object(2)
memory usage: 273.6+ KB
None
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3500 entries, 0 to 3499
Data columns (total 2 columns):
 #   Column   Non-Null Count  Dtype
---  ------   --------------  -----
 0   cust_id  3500 non-null   int64
 1   gender   3500 non-null   int64
dtypes: int64(2)
memory usage: 54.8 KB
None



데이터의 요약 정보나 통계 정보를 출력해 변수들의 유형과 분포를 확인합니다.

In [6]:
print(X.describe())  
print(y.describe())


           cust_id          총구매액         최대구매액          환불금액         내점일수  \
count  3500.000000  3.500000e+03  3.500000e+03  1.205000e+03  3500.000000   
mean   1749.500000  9.191925e+07  1.966424e+07  2.407822e+07    19.253714   
std    1010.507298  1.635065e+08  3.199235e+07  4.746453e+07    27.174942   
min       0.000000 -5.242152e+07 -2.992000e+06  5.600000e+03     1.000000   
25%     874.750000  4.747050e+06  2.875000e+06  2.259000e+06     2.000000   
50%    1749.500000  2.822270e+07  9.837000e+06  7.392000e+06     8.000000   
75%    2624.250000  1.065079e+08  2.296250e+07  2.412000e+07    25.000000   
max    3499.000000  2.323180e+09  7.066290e+08  5.637530e+08   285.000000   

           내점당구매건수       주말방문비율         구매주기  
count  3500.000000  3500.000000  3500.000000  
mean      2.834963     0.307246    20.958286  
std       1.912368     0.289752    24.748682  
min       1.000000     0.000000     0.000000  
25%       1.666667     0.027291     4.000000  
50%       2.333333     0

<br/>

---

<br/>
<br/>

# 3. 데이터 전처리
* 전처리 과정을 통해서 머신러닝에 사용할 수 있는 형태의 데이터 준비


필요한 라이브러리를 불러옵니다.
- 인코딩 : LabelEncoder
- 데이터 표준화 : StandardScaler

In [7]:
from sklearn.preprocessing import LabelEncoder
from sklearn.preprocessing import StandardScaler

* 단순히 1부터의 숫자를 부여한 'cust_id'를 수치형 변수로 받아들이면, 결과가 왜곡될 수 있으니 컬럼을 제거합니다.

In [8]:
# X = X.drop(columns=['cust_id'])
# y = y.drop(columns=['cust_id'])

In [9]:
print(X.info())
print(y.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3500 entries, 0 to 3499
Data columns (total 10 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   cust_id  3500 non-null   int64  
 1   총구매액     3500 non-null   int64  
 2   최대구매액    3500 non-null   int64  
 3   환불금액     1205 non-null   float64
 4   주구매상품    3500 non-null   object 
 5   주구매지점    3500 non-null   object 
 6   내점일수     3500 non-null   int64  
 7   내점당구매건수  3500 non-null   float64
 8   주말방문비율   3500 non-null   float64
 9   구매주기     3500 non-null   int64  
dtypes: float64(3), int64(5), object(2)
memory usage: 273.6+ KB
None
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3500 entries, 0 to 3499
Data columns (total 2 columns):
 #   Column   Non-Null Count  Dtype
---  ------   --------------  -----
 0   cust_id  3500 non-null   int64
 1   gender   3500 non-null   int64
dtypes: int64(2)
memory usage: 54.8 KB
None


- 데이터에 결측치가 있는지 확인해보세요


In [10]:
print(X.isnull().sum()) # 환불금액 결측치 2295건 발생

cust_id       0
총구매액          0
최대구매액         0
환불금액       2295
주구매상품         0
주구매지점         0
내점일수          0
내점당구매건수       0
주말방문비율        0
구매주기          0
dtype: int64


- 결측치에 0으로 채워 넣어 모델 학습에 지장이 없도록 합니다.

In [11]:
# cust_id 기준으로 병합
df = pd.merge(X, y, on='cust_id', how='inner')
df['환불금액'] = np.where(
    (df['총구매액'] == 0) & (df['환불금액'].isna()),
    0,
    df['환불금액']
)

# 1. 환불금액이 NaN이 아닌 행만 사용해 환불률 계산
valid_refund_df = df[df['환불금액'].notna()].copy()

# 총구매액이 0인 경우는 제외 (inf 방지)
valid_refund_df = valid_refund_df[valid_refund_df['총구매액'] != 0]

# 2. 성별별 평균 환불률 계산
valid_refund_df['환불률'] = valid_refund_df['환불금액'] / valid_refund_df['총구매액']
avg_refund_rate_by_gender = valid_refund_df.groupby('gender')['환불률'].mean()

# 3. NaN 환불금액을 성별 평균 환불률로 대체
def fill_missing_refund(row):
    if pd.isna(row['환불금액']):
        refund_rate = avg_refund_rate_by_gender.loc[row['gender']]
        return row['총구매액'] * refund_rate
    else:
        return row['환불금액']

df['환불금액'] = df.apply(fill_missing_refund, axis=1)

# y: 성별(예측 대상)
y = df['gender']

# X: 성별 컬럼 제외한 나머지
X = df.drop(columns=['gender'])


In [12]:
print(X.isnull().sum())

cust_id    0
총구매액       0
최대구매액      0
환불금액       0
주구매상품      0
주구매지점      0
내점일수       0
내점당구매건수    0
주말방문비율     0
구매주기       0
dtype: int64


In [13]:
X = X.drop(columns=['cust_id'])
y = y.drop(columns=['cust_id'])

문자형 범주 데이터를 숫자로 바꾸기 위한 인코딩을 수행합니다.

In [14]:
le_1 = LabelEncoder()
le_2= LabelEncoder()

# 문자형 → 숫자형으로 변환
X['주구매상품'] = le_1.fit_transform(X['주구매상품'])
X['주구매지점'] = le_2.fit_transform(X['주구매지점'])


In [15]:
print(X[['주구매상품', '주구매지점']].head())

   주구매상품  주구매지점
0      5      0
1     21     19
2      6      1
3      5      2
4     15      8


In [16]:
print(X.info())
print(y.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3500 entries, 0 to 3499
Data columns (total 9 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   총구매액     3500 non-null   int64  
 1   최대구매액    3500 non-null   int64  
 2   환불금액     3500 non-null   float64
 3   주구매상품    3500 non-null   int64  
 4   주구매지점    3500 non-null   int64  
 5   내점일수     3500 non-null   int64  
 6   내점당구매건수  3500 non-null   float64
 7   주말방문비율   3500 non-null   float64
 8   구매주기     3500 non-null   int64  
dtypes: float64(3), int64(6)
memory usage: 246.2 KB
None
<class 'pandas.core.series.Series'>
RangeIndex: 3500 entries, 0 to 3499
Series name: gender
Non-Null Count  Dtype
--------------  -----
3500 non-null   int64
dtypes: int64(1)
memory usage: 27.5 KB
None


각 데이터에 표준화를 적용하여 데이터의 스케일(크기 차이)을 맞춰줍니다.
- 평균을 0, 표준편차를 1로 맞춰서 → 데이터가 정규 분포 형태로 변환되도록 하세요

In [17]:
scaler = StandardScaler()

# 수치형 컬럼만 추출- ( fit_transform() 반환: numpy, select_dtypes() 반환:pandas )
numeric_cols = X.select_dtypes(include=['int64', 'float64']).columns

# 표준화 적용 - 모델이 학습할때 숫자의 크기와 무관하게 특성 자체의 기여도를 측정하기 위함
X[numeric_cols] = scaler.fit_transform(X[numeric_cols])


In [18]:
X.head()

,총구매액,최대구매액,환불금액,주구매상품,주구매지점,내점일수,내점당구매건수,주말방문비율,구매주기
0,-0.144580,-0.262608,-0.363646,-0.738357,-1.904703,-0.009338,0.554247,0.758623,-0.159962
1,-0.549190,-0.547967,-0.469981,0.490702,1.466677,-0.635003,-0.698168,-1.060530,-0.806554
2,-0.542700,-0.563504,-0.452309,-0.661541,-1.727262,-0.635003,-0.436675,-1.060530,-0.806554
3,-0.463911,-0.460465,-0.361518,-0.738357,-1.549821,-0.046142,-0.204236,0.037746,-0.200374
4,-0.384561,0.135544,-0.066460,0.029805,-0.485175,-0.635003,-0.698168,-1.060530,2.588052


In [19]:
print(X.describe())

               총구매액         최대구매액          환불금액         주구매상품         주구매지점  \
count  3.500000e+03  3.500000e+03  3.500000e+03  3.500000e+03  3.500000e+03   
mean  -3.349701e-17 -4.263256e-17  6.090366e-17 -6.090366e-17 -8.767590e-17   
std    1.000143e+00  1.000143e+00  1.000143e+00  1.000143e+00  1.000143e+00   
min   -8.829091e-01 -7.082781e-01 -4.748082e-01 -1.122438e+00 -1.904703e+00   
25%   -5.332182e-01 -5.248643e-01 -4.478145e-01 -7.383573e-01 -4.851745e-01   
50%   -3.896215e-01 -3.072187e-01 -3.566322e-01 -4.310925e-01 -3.077335e-01   
75%    8.923657e-02  1.031100e-01 -1.449802e-02  5.675181e-01  7.569129e-01   
max    1.364826e+01  2.147585e+01  1.385126e+01  2.027026e+00  2.176441e+00   

               내점일수       내점당구매건수        주말방문비율          구매주기  
count  3.500000e+03  3.500000e+03  3.500000e+03  3.500000e+03  
mean   4.669281e-17 -2.618858e-16  7.917476e-17 -4.263256e-17  
std    1.000143e+00  1.000143e+00  1.000143e+00  1.000143e+00  
min   -6.718072e-01 -9.596611e-0

<br/>

---

<br/>
<br/>

# 5-1. 모델링 - LogisticRegression

* 본격적으로 모델을 선언하고 학습시킵니다.


필요한 라이브러리를 불러옵니다.

In [20]:
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler


모델을 선언하여 객체화시킵니다.

In [21]:
lr_clf = LogisticRegression(solver='liblinear', random_state=42)
lr_clf


LogisticRegression(random_state=42, solver='liblinear')

모델을 학습 데이터에 맞춰 학습시킵니다.

In [22]:
# 데이터 분할
X_train,X_test, y_train,y_test = train_test_split(X, y, test_size= 0.2, random_state=42)

In [23]:
# 스케일링
scaler1 = StandardScaler()
scaler1.fit(X_train)
X_train_scaled = scaler1.transform(X_train)
X_test_scaled = scaler1.transform(X_test)

In [24]:
from sklearn.model_selection import GridSearchCV
param_grid_lr = {
    'penalty': ['l1', 'l2'],
    'C': [0.01, 0.1, 1, 10],
    'solver': ['liblinear']  # 'liblinear'은 l1, l2 모두 가능
}

grid_search_lr = GridSearchCV(
    estimator=lr_clf,
    param_grid=param_grid_lr,
    scoring='accuracy',    # 정확도로 평가
    cv=5,                  # 5-fold 교차검증
    n_jobs=-1              # 모든 코어 사용
)

# 모델 학습
grid_search_lr.fit(X_train_scaled, y_train)
print("Best Parameters:", grid_search_lr.best_params_)

# GridSearchCV로 찾은 최적 모델 추출
best_model = grid_search_lr.best_estimator_

Best Parameters: {'C': 0.1, 'penalty': 'l1', 'solver': 'liblinear'}


<br/>

---

<br/>
<br/>

# 6-1. 예측 성능 확인해보기 - LogisticRegression

- 학습된 모델로 테스트 데이터에 대한 예측을 수행합니다.

In [25]:
# 테스트 데이터로 예측
pred_lr = best_model.predict(X_test)

- 학습시킨 모델의 성능을 알아봅니다
- 각 평가지표로 모델의 성능을 수치화하여 확인합니다.
- 필요한 라이브러리를 import 하고 성능을 확인해보세요 (정확도, 정밀도, 재현율, f1, confusion_matrix)

In [26]:
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, confusion_matrix, classification_report
# 정확도
acc = accuracy_score(y_test, pred_lr)
# 정밀도
prec = precision_score(y_test, pred_lr)
# 재현율
rec = recall_score(y_test, pred_lr)
# F1 점수
f1 = f1_score(y_test, pred_lr)
# 혼동 행렬
cm = confusion_matrix(y_test, pred_lr)


In [27]:
print(f"Accuracy (정확도): {acc:.4f}")
print(f"Precision (정밀도): {prec:.4f}")
print(f"Recall (재현율): {rec:.4f}")
print(f"F1 Score: {f1:.4f}")
print("Confusion Matrix (혼동 행렬):")
print(cm)


Accuracy (정확도): 0.6171
Precision (정밀도): 0.5410
Recall (재현율): 0.1209
F1 Score: 0.1976
Confusion Matrix (혼동 행렬):
[[399  28]
 [240  33]]


In [28]:
print(classification_report(y_test, pred_lr))


              precision    recall  f1-score   support

           0       0.62      0.93      0.75       427
           1       0.54      0.12      0.20       273

    accuracy                           0.62       700
   macro avg       0.58      0.53      0.47       700
weighted avg       0.59      0.62      0.53       700




<br/>

---

<br/>
<br/>

# 5-2. 모델링 - DecisionTreeClassifier

* 본격적으로 모델을 선언하고 학습시킵니다.


필요한 라이브러리를 불러옵니다.

In [29]:
from sklearn.tree import DecisionTreeClassifier

모델을 선언하여 객체화시킵니다.

In [30]:
dt_clf = DecisionTreeClassifier()
dt_clf

DecisionTreeClassifier()

모델을 학습 데이터에 맞춰 학습시킵니다.

In [31]:
# 모델학습
dt_clf.fit(X_train, y_train)
dt_clf

DecisionTreeClassifier()



<br/>
<br/>

# 6-2. 예측 성능 확인해보기 - DecisionTreeClassifier

- 학습된 모델로 테스트 데이터에 대한 예측을 수행합니다.

In [32]:
## 모델 예측
dpred = dt_clf.predict(X_test)
dpred

array([1, 1, 0, 0, 0, 1, 1, 0, 0, 0, 0, 1, 0, 1, 1, 1, 1, 0, 0, 0, 0, 0,
       1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 1, 1, 0, 0, 1, 1, 0,
       1, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 1, 1, 0, 0, 0, 0, 0,
       0, 0, 0, 1, 0, 1, 0, 1, 0, 0, 1, 0, 0, 0, 1, 1, 1, 1, 1, 0, 0, 0,
       1, 1, 1, 0, 0, 0, 0, 0, 1, 0, 1, 1, 0, 0, 0, 0, 0, 1, 1, 1, 1, 0,
       0, 1, 1, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0, 1, 0, 1,
       1, 1, 1, 0, 0, 0, 1, 0, 1, 0, 0, 1, 1, 0, 1, 0, 0, 0, 0, 0, 1, 0,
       0, 1, 1, 0, 0, 0, 1, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0,
       0, 1, 0, 1, 1, 1, 1, 1, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 1, 1, 1, 0,
       0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 1, 0, 0, 0, 1, 1, 0, 0, 0, 1, 0, 0,
       0, 1, 0, 0, 1, 0, 1, 1, 0, 0, 1, 0, 0, 1, 0, 1, 1, 0, 1, 1, 0, 0,
       1, 0, 1, 0, 0, 0, 1, 0, 1, 0, 1, 0, 0, 1, 0, 0, 0, 1, 0, 1, 0, 0,
       0, 1, 0, 1, 1, 0, 1, 0, 0, 1, 0, 1, 0, 0, 1, 0, 0, 1, 0, 1, 0, 0,
       0, 0, 1, 1, 0, 0, 0, 0, 0, 1, 0, 1, 1, 0, 0,

- 학습시킨 모델의 성능을 알아봅니다
- 각 평가지표로 모델의 성능을 수치화하여 확인합니다.
- 필요한 라이브러리를 import 하고 성능을 확인해보세요 (정확도, 정밀도, 재현율, f1, confusion_matrix)

In [33]:
##평가
accuracy_score(y_test, dpred)

0.8228571428571428

In [34]:
# 정확도
acc = accuracy_score(y_test, dpred)
# 정밀도
prec = precision_score(y_test, dpred)
# 재현율
rec = recall_score(y_test, dpred)
# F1 점수
f1 = f1_score(y_test, dpred)
# 혼동 행렬
cm = confusion_matrix(y_test, dpred)

In [35]:
print(f"Accuracy (정확도): {acc:.4f}")
print(f"Precision (정밀도): {prec:.4f}")
print(f"Recall (재현율): {rec:.4f}")
print(f"F1 Score: {f1:.4f}")
print("Confusion Matrix (혼동 행렬):")
print(cm)


Accuracy (정확도): 0.8229
Precision (정밀도): 0.7854
Recall (재현율): 0.7509
F1 Score: 0.7678
Confusion Matrix (혼동 행렬):
[[371  56]
 [ 68 205]]


In [36]:
print(classification_report(y_test, dpred))


              precision    recall  f1-score   support

           0       0.85      0.87      0.86       427
           1       0.79      0.75      0.77       273

    accuracy                           0.82       700
   macro avg       0.82      0.81      0.81       700
weighted avg       0.82      0.82      0.82       700




<br/>

---

<br/>
<br/>

# 5-3. 모델링 - RandomForestClassifier

* 본격적으로 모델을 선언하고 학습시킵니다.



필요한 라이브러리를 불러옵니다.

In [37]:
from sklearn.ensemble import RandomForestClassifier

모델을 선언하여 객체화시킵니다.

In [38]:
rf_clf = RandomForestClassifier(random_state=42)
rf_clf

RandomForestClassifier(random_state=42)

모델을 학습 데이터에 맞춰 학습시킵니다.

In [39]:
rf_clf.fit(X_train, y_train)

RandomForestClassifier(random_state=42)



<br/>
<br/>

# 6-3. 예측 성능 확인해보기 - RandomForestClassifier

- 학습된 모델로 테스트 데이터에 대한 예측을 수행합니다.

In [40]:
pred_rf = rf_clf.predict(X_test)

- 학습시킨 모델의 성능을 알아봅니다
- 각 평가지표로 모델의 성능을 수치화하여 확인합니다.
- 필요한 라이브러리를 import 하고 성능을 확인해보세요 (정확도, 정밀도, 재현율, f1, confusion_matrix)

In [41]:
##평가
accuracy_score(y_test, pred_rf)

0.8342857142857143

In [42]:
# 정확도
acc = accuracy_score(y_test, pred_rf)
# 정밀도
prec = precision_score(y_test, pred_rf)
# 재현율
rec = recall_score(y_test, pred_rf)
# F1 점수
f1 = f1_score(y_test, pred_rf)
# 혼동 행렬
cm = confusion_matrix(y_test, pred_rf)


In [43]:
print(classification_report(y_test, pred_rf))

              precision    recall  f1-score   support

           0       0.83      0.92      0.87       427
           1       0.85      0.70      0.77       273

    accuracy                           0.83       700
   macro avg       0.84      0.81      0.82       700
weighted avg       0.84      0.83      0.83       700




<br/>

---

<br/>
<br/>

# 5-4. 모델링 - XGBoost

* 본격적으로 모델을 선언하고 학습시킵니다.



필요한 라이브러리를 불러옵니다.

In [44]:
import xgboost as xgb
print(xgb.__version__)

1.6.2


모델을 선언하여 객체화시킵니다.

In [45]:
# 모델 생성
xgb_clf = xgb.XGBClassifier(random_state=42, use_label_encoder=False, eval_metric='logloss')
xgb_clf

XGBClassifier(base_score=None, booster=None, callbacks=None,
              colsample_bylevel=None, colsample_bynode=None,
              colsample_bytree=None, early_stopping_rounds=None,
              enable_categorical=False, eval_metric='logloss', gamma=None,
              gpu_id=None, grow_policy=None, importance_type=None,
              interaction_constraints=None, learning_rate=None, max_bin=None,
              max_cat_to_onehot=None, max_delta_step=None, max_depth=None,
              max_leaves=None, min_child_weight=None, missing=nan,
              monotone_constraints=None, n_estimators=100, n_jobs=None,
              num_parallel_tree=None, predictor=None, random_state=42,
              reg_alpha=None, reg_lambda=None, ...)

모델을 학습 데이터에 맞춰 학습시킵니다.

In [46]:
# 학습
xgb_clf.fit(X_train, y_train)

XGBClassifier(base_score=0.5, booster='gbtree', callbacks=None,
              colsample_bylevel=1, colsample_bynode=1, colsample_bytree=1,
              early_stopping_rounds=None, enable_categorical=False,
              eval_metric='logloss', gamma=0, gpu_id=-1,
              grow_policy='depthwise', importance_type=None,
              interaction_constraints='', learning_rate=0.300000012,
              max_bin=256, max_cat_to_onehot=4, max_delta_step=0, max_depth=6,
              max_leaves=0, min_child_weight=1, missing=nan,
              monotone_constraints='()', n_estimators=100, n_jobs=0,
              num_parallel_tree=1, predictor='auto', random_state=42,
              reg_alpha=0, reg_lambda=1, ...)



<br/>
<br/>

# 6-4. 예측 성능 확인해보기 - XGBoost

- 학습된 모델로 테스트 데이터에 대한 예측을 수행합니다.

In [47]:
# 예측
pred_xgb = xgb_clf.predict(X_test)

- 학습시킨 모델의 성능을 알아봅니다
- 각 평가지표로 모델의 성능을 수치화하여 확인합니다.
- 필요한 라이브러리를 import 하고 성능을 확인해보세요 (정확도, 정밀도, 재현율, f1, confusion_matrix)

In [48]:
##평가
accuracy_score(y_test, pred_xgb)


0.8714285714285714

In [49]:
# 정확도
acc = accuracy_score(y_test, dpred)
# 정밀도
prec = precision_score(y_test, dpred)
# 재현율
rec = recall_score(y_test, dpred)
# F1 점수
f1 = f1_score(y_test, dpred)
# 혼동 행렬
cm = confusion_matrix(y_test, dpred)

In [50]:
print(classification_report(y_test, pred_lr))

              precision    recall  f1-score   support

           0       0.62      0.93      0.75       427
           1       0.54      0.12      0.20       273

    accuracy                           0.62       700
   macro avg       0.58      0.53      0.47       700
weighted avg       0.59      0.62      0.53       700



<br/>

---


<br/>

## 7.  위 4가지 모델의 학습 & 예측 & 평가 결과를 확인하고 최고 성능을 내는 모델을 찾아봅시다!

- 어떤 모델이 가장 성능이 좋은가요 ?

| 모델명                    | Accuracy | Precision (avg)  | Recall (avg)     | F1-score (avg)   |
| ---------------------- | -------- | ---------------- | ---------------- | ---------------- |
| **LogisticRegression** | 0.62     | 0.58 (macro)     | 0.53 (macro)     | 0.47 (macro)     |
| **DecisionTree**       | 0.82     | 0.82 (macro)     | 0.81 (macro)     | 0.81 (macro)     |
| **RandomForest**       | **0.83** | **0.84 (macro)** | **0.81 (macro)** | **0.82 (macro)** |
| **XGBoost**            | 0.62     | 0.58 (macro)     | 0.53 (macro)     | 0.47 (macro)     |


결론: **RandomForestClassifier**가 가장 우수한 모델입니다

이유:
전체 **Accuracy**가 가장 높음 (**0.83**)

두 클래스 모두에서 Precision, Recall, F1-score가 고르게 우수

특히, 불균형 데이터로 보이는 상황에서 class 1의 Recall=0.70, F1=0.77로 꽤 잘 잡아내고 있음

DecisionTree도 좋지만, RandomForest가 약간 더 높은 정밀성과 안정성을 보임